### PR TITLE
crimson/osd/backfill_state: fixes two corner cases in backfilling

### PR DESCRIPTION
**Pull request by xxhdx1985126**
1. BACKFILL_FINISH messages may be dropped by replicas, if osdmap updates get processed by replicas before BACKFILL_FINISH messages; in this case, the peering for the osdmap updates will find replicas' backfills not finished, and try to do another round of backfilling, but, since there's nothing to be backfilled, there's no way to update peer backfill infos, which means the pg may never go clean.
2. The last round of scan may produce no objects that need to be backfilled; in this case, the current implementation doesn't set backfill infos for the primary and replicas and leaves the pg as unclean forever.

Signed-off-by: Xuehan Xu <xuxuehan@qianxin.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [x] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by athanatos** [comment]
@rzarzynski Can you take a look at this one?  I'm not sure I'm following what the invariants for ReplicasScanning/PrimaryScanning/Enqueueing are -- looks like pushes can be completed in any of them?

**Comment by athanatos** [comment]
@rzarzynski ping

**Comment by xxhdx1985126** [comment]
@rzarzynski Could you take a look at this PR, please? Thanks:-)

**Comment by xxhdx1985126** [comment]
@rzarzynski please take a look, thanks:-)

**Comment by xxhdx1985126** [comment]
jenkins test make check

File: src/crimson/osd/backfill_state.cc [hunk 1]
```diff
@@ -352,9 +352,14 @@ BackfillState::Enqueuing::Enqueuing(my_context ctx)
     logger().debug("{}: reached end for current local chunk",
                    __func__);
     post_event(RequestPrimaryScanning{});
-  } else if (backfill_state().progress_tracker->tracked_objects_completed()) {
-    post_event(RequestDone{});
```
Comment (rzarzynski) [inline src/crimson/osd/backfill_state.cc:354]: I think the reaction for `RequestDone` has been removed from `Enqueuing` in the previous commit but I see the reason for splitting the work between two patches. OK.

**Review by rzarzynski — APPROVED** [review]
LGTM!

**Comment by Matan-B** [comment]
https://shaman.ceph.com/builds/ceph/wip-matanb-crimson-testing-aug-12

**Comment by Matan-B** [comment]
https://pulpito.ceph.com/matan-2024-08-14_16:27:16-crimson-rados-wip-matanb-crimson-testing-aug-14-distro-crimson-smithi/

7855673, 7855704 - infra
7855680 - new, untreated https://tracker.ceph.com/issues/67568
7855707 - https://tracker.ceph.com/issues/66767

**Pull request merged**